File: library/python/src/pywavemap/convert.py
"""
convert
*******
Submodule with common conversion functions for wavemap index types.
"""

import numpy as _np
from ._pywavemap_bindings import OctreeIndex


def cell_width_to_height(cell_width, min_cell_width):
    """
    Compute the minimum node height (resolution level) required to reach
    a given width.
    """
    return int(_np.ceil(_np.log2(cell_width / min_cell_width)))


def height_to_cell_width(min_cell_width, height):
    """Compute the node width at a given height."""
    return min_cell_width * float(_np.exp2(height))


def scaled_point_to_nearest_index(point):
    """Compute the nearest index to a point on the unit grid."""
    return (point - 0.5).round().astype(int)


def point_to_nearest_index(point, cell_width):
    """
    Compute the nearest index to a point on a grid with a given cell width.
    """
    return scaled_point_to_nearest_index(point / cell_width)


def point_to_node_index(point, min_cell_width, height):
    """
    Compute the index of a node containing a given point.

    :param min_cell_width: The grid resolution at height 0 (max map resolution).
    :param height: The desired height (resolution level) of the node index.
    """
    node_width = height_to_cell_width(min_cell_width, height)
    position_index = point_to_nearest_index(point, node_width)
    return OctreeIndex(height, position_index)

File: library/python/src/convert.cc
#include "pywavemap/convert.h"

#include <nanobind/eigen/dense.h>
#include <wavemap/core/indexing/index_conversions.h>

using namespace nb::literals;  // NOLINT

namespace wavemap {
void add_convert_module(nb::module_& m_convert) {
  m_convert.def(
      "cell_width_to_height",
      [](FloatingPoint cell_width, FloatingPoint min_cell_width) {
        return convert::cellWidthToHeight(cell_width, 1.f / min_cell_width);
      },
      "cell_width"_a, "min_cell_width"_a,
      "Compute the minimum node height (resolution level) required to reach"
      "a given node width.\n\n"
      "    :param cell_width: The desired node width.\n"
      "    :param min_cell_width: The grid resolution at height 0 "
      "(max map resolution).");

  m_convert.def("height_to_cell_width", &convert::heightToCellWidth,
                "min_cell_width"_a, "height"_a,
                "Compute the node width at a given height.\n\n"
                "    :param min_cell_width: The grid resolution at height 0 "
                "(max map resolution).\n"
                "    :param height: The desired height (resolution level) of "
                "the node index.");

  m_convert.def(
      "point_to_nearest_index",
      [](const Point3D& point, FloatingPoint cell_width) {
        return convert::pointToNearestIndex<3>(point, 1.f / cell_width);
      },
      "point"_a, "cell_width"_a,
      "Compute the nearest index to a point on a grid with a given "
      "cell width.");

  m_convert.def("point_to_node_index", &convert::pointToNodeIndex<3>, "point"_a,
                "min_cell_width"_a, "height"_a,
                "Compute the index of a node containing a given point.\n\n"
                "    :param min_cell_width: The grid resolution at height 0 "
                "(max map resolution).\n"
                "    :param height: The desired height (resolution level) of "
                "the node index.");
}
}  // namespace wavemap
